main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala

```diff
@@ -298,31 +303,31 @@ object ErasedAst {
 
   object ConstraintParam {
 
-    case class HeadParam(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.ConstraintParam
+    case class HeadParam(sym: Symbol.VarSym, tpe: ErasedType[JType], loc: SourceLocation) extends ErasedAst.ConstraintParam
 
-    case class RuleParam(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.ConstraintParam
+    case class RuleParam(sym: Symbol.VarSym, tpe: ErasedType[JType], loc: SourceLocation) extends ErasedAst.ConstraintParam
 
   }
 
-  case class FormalParam(sym: Symbol.VarSym, tpe: MonoType)
+  case class FormalParam(sym: Symbol.VarSym, tpe: ErasedType[JType])
 
-  case class FreeVar(sym: Symbol.VarSym, tpe: MonoType)
+  case class FreeVar(sym: Symbol.VarSym, tpe: ErasedType[JType])
 
-  case class BoxInt8(exp: ErasedAst.Expression[PrimInt8], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedInt8]
-  case class BoxInt16(exp: ErasedAst.Expression[PrimInt16], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedInt16]
-  case class BoxInt32(exp: ErasedAst.Expression[PrimInt32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedInt32]
-  case class BoxInt64(exp: ErasedAst.Expression[PrimInt64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedInt64]
-  case class BoxChar(exp: ErasedAst.Expression[PrimChar], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedChar]
-  case class BoxFloat32(exp: ErasedAst.Expression[PrimFloat32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedFloat32]
-  case class BoxFloat64(exp: ErasedAst.Expression[PrimFloat64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedFloat64]
+  case class BoxInt8(exp: ErasedAst.Expression[PrimInt8], final val tpe: ErasedType[BoxedInt8], loc: SourceLocation) extends ErasedAst.Expression[BoxedInt8]
```

Comment:
What I intended to have was:

```
case class BoxInt8(exp: ErasedAst.Expression[PrimInt8], loc: SourceLocation) {
  final val tpe = ErasedType.BoxInt8[PrimInt8]
}
```

makes sense? This applies both for the boxing and unboxing operations.

Reply:
Was this new version what you meant?

Reply:
Yes